File: sample-chat-obj-c/sample-chat/Components/DialogsScreen/DialogsSelectionViewController.m
//
//  DialogsSelectionVC.m
//  sample-chat
//
//  Created by Injoit on 1/31/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "DialogsSelectionViewController.h"
#import "QBUUser+Chat.h"
#import "TitleView.h"
#import "UIView+Chat.h"
#import "UIViewController+Alert.h"

NSString *const MESSAGE_FORWARDED = @"Message forwarded";
NSString *const DELETE_CHATS = @"Delete Chats";
NSString *const KEY_MESSAGE_FORWARDED = @"origin_sender_name";

@interface DialogsSelectionViewController()
//MARK: - Properties
@property (nonatomic, strong) TitleView *titleView;
@property (nonatomic, strong) NSMutableSet<NSIndexPath *> *selectedPaths;
@end

@implementation DialogsSelectionViewController

- (void)setupDialogs {
    [self reloadContent];
    self.selectedPaths = [NSMutableSet set];
}

- (void)setupViews {
    self.titleView = [[TitleView alloc] init];
    self.navigationItem.titleView = self.titleView;
    UIBarButtonItem *backButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"chevron"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(didTapBack:)];
    backButtonItem.tintColor = UIColor.whiteColor;
    self.navigationItem.leftBarButtonItem = backButtonItem;
    if (self.action == ChatActionDelete) {
        UIBarButtonItem *deleteButtonItem = [[UIBarButtonItem alloc] initWithTitle: @"Delete"
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(didTapDelete:)];
        self.navigationItem.rightBarButtonItem = deleteButtonItem;
        deleteButtonItem.tintColor = UIColor.whiteColor;
    } else if (self.action == ChatActionForward)  {
        UIBarButtonItem *sendButtonItem = [[UIBarButtonItem alloc] initWithTitle: @"Send"
                                                                           style:UIBarButtonItemStylePlain
                                                                          target:self
                                                                          action:@selector(didTapSend:)];
        self.navigationItem.rightBarButtonItem = sendButtonItem;
        sendButtonItem.tintColor = UIColor.whiteColor;
    }
    self.navigationItem.rightBarButtonItem.enabled = NO;
}

- (void)setupNavigationTitle {
    NSString *title = self.action == ChatActionDelete ? DELETE_CHATS : MESSAGE_FORWARDED;
    NSString *chats = self.selectedPaths.count > 1 ? @"chats" : @"chat";
    NSString *numberChats = [NSString stringWithFormat:@"%@ %@ selected", @(self.selectedPaths.count), chats];
    [self.titleView setupTitleViewWithTitle:title subTitle:numberChats];
}

#pragma mark - Actions
- (void)didTapSend:(UIButton *)sender {
    if (QBChat.instance.isConnected == NO) {
        [self showNoInternetAlertWithHandler:nil];
        return;
    }
    if (!self.message) {
        return;
    }
    Profile *currentUser = [[Profile alloc] init];
    if (currentUser.isFull == NO) {
        return;
    }
    sender.enabled = NO;
    [self.progressView start];
    dispatch_group_t sendGroup = dispatch_group_create();
    for (NSIndexPath *indexPath in self.selectedPaths) {
        QBChatDialog *dialog = self.dialogs[indexPath.row];
        if (!dialog) {
            continue;
        }
        dispatch_group_enter(sendGroup);
        QBChatMessage *forwardedMessage = [QBChatMessage markableMessage];
        forwardedMessage.senderID = currentUser.ID;
        forwardedMessage.dialogID = dialog.ID;
        forwardedMessage.dateSent = [NSDate date];
        forwardedMessage.deliveredIDs = @[@(currentUser.ID)];
        forwardedMessage.readIDs = @[@(currentUser.ID)];
        forwardedMessage.customParameters[@"save_to_history"] = @"1";
        forwardedMessage.markable = YES;
        QBUUser *originSenderUser = [self.chatManager.storage userWithID:self.message.senderID];
        if (originSenderUser){
            forwardedMessage.customParameters[KEY_MESSAGE_FORWARDED] = originSenderUser.fullName;
        } else {
            forwardedMessage.customParameters[KEY_MESSAGE_FORWARDED] = currentUser.fullName;
        }
        if (self.message.attachments.firstObject) {
            QBChatAttachment *attachment = self.message.attachments.firstObject;
            forwardedMessage.text = @"[Attachment]";
            forwardedMessage.attachments = @[attachment];
        } else {
            forwardedMessage.text = self.message.text;
        }
        [self.chatManager sendMessage:forwardedMessage toDialog:dialog completion:^(NSError * _Nullable error) {
            dispatch_group_leave(sendGroup);
            if (error) {
                Log(@"[%@] Send Forwarded Message error: %@",  NSStringFromClass([DialogsSelectionViewController class]), error.localizedDescription);
            }
            if (!error) {
                [self.selectedPaths removeObject:indexPath];
            }
        }];
    }
    dispatch_group_notify(sendGroup, dispatch_get_main_queue(), ^{
        [self handleLeaveDialog];
    });
}

- (void)didTapBack:(UIButton *)sender {
    [self.navigationController popViewControllerAnimated:NO];
}

- (void)didTapDelete:(UIButton *)sender {
    if (QBChat.instance.isConnected == NO) {
        [self showNoInternetAlertWithHandler:nil];
        return;
    }
    NSString *baseAlertMessage = @"Do you really want to leave selected dialog";
    NSString *alertMessage = self.selectedPaths.count == 1 ? [baseAlertMessage stringByAppendingString:@"?"] : [baseAlertMessage stringByAppendingString:@"s?"];
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Warning" message:alertMessage preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:nil];
    UIAlertAction *leaveAction = [UIAlertAction actionWithTitle:@"Leave"
                                                          style:UIAlertActionStyleDefault
                                                        handler:^(UIAlertAction * _Nonnull action) {
        sender.enabled = NO;
        [self.progressView start];
        dispatch_group_t deleteGroup = dispatch_group_create();
        for (NSIndexPath *indexPath in self.selectedPaths) {
            QBChatDialog *dialog = self.dialogs[indexPath.row];
            if (dialog.type == QBChatDialogTypePublicGroup) {
                continue;
            }
            dispatch_group_enter(deleteGroup);
            [self.chatManager leaveDialogWithID:dialog.ID completion:^(NSString * _Nullable error) {
                dispatch_group_leave(deleteGroup);
                if (error) {
                    Log(@"[%@] Leave Dialog error: %@",  NSStringFromClass([DialogsSelectionViewController class]), error);
                }
                if (!error) {
                    [self.selectedPaths removeObject:indexPath];
                }
            }];
        }
        dispatch_group_notify(deleteGroup, dispatch_get_main_queue(), ^{
            [self handleLeaveDialog];
        });
    }];
    
    [alertController addAction:cancelAction];
    [alertController addAction:leaveAction];
    [self presentViewController:alertController animated:NO completion:nil];
}

#pragma mark - Helpers
- (void)checkNavRightBarButtonEnabled {
    self.navigationItem.rightBarButtonItem.enabled = self.selectedPaths.count > 0;
}

- (void)reloadContent {
    self.dialogs = [[ChatManager instance].storage dialogsSortByUpdatedAt];
    [self setupNavigationTitle];
    [self.tableView reloadData];
}

- (void)handleLeaveDialog {
    [self.progressView stop];
    [self.navigationController popViewControllerAnimated:NO];
}

- (void)configureCell:(DialogCell *)cell forIndexPath:(NSIndexPath *)indexPath {
    self.tableView.allowsMultipleSelection = YES;
    cell.checkBoxView.hidden = NO;
    cell.unreadMessageCounterLabel.hidden = YES;
    cell.unreadMessageCounterHolder.hidden = YES;
    cell.lastMessageDateLabel.hidden = YES;
    if ([self.selectedPaths containsObject:indexPath]) {
        cell.contentView.backgroundColor = [UIColor colorWithRed:0.85f green:0.89f blue:0.97f alpha:1.0f];
        cell.checkBoxImageView.hidden = NO;
        [cell.checkBoxView setRoundBorderEdgeColorView:4.0f
                                           borderWidth:0.0f
                                                 color:[UIColor colorWithRed:0.22f green:0.47f blue:0.99f alpha:1.0f]
                                           borderColor:[UIColor colorWithRed:0.22f green:0.47f blue:0.99f alpha:1.0f]];
    } else {
        cell.contentView.backgroundColor = UIColor.clearColor;
        cell.checkBoxView.backgroundColor = UIColor.clearColor;
        [cell.checkBoxView setRoundBorderEdgeColorView:4.0f
                                           borderWidth:1.0f
                                                 color:nil
                                           borderColor:[UIColor colorWithRed:0.42f green:0.48f blue:0.57f alpha:1.0f]];
        cell.checkBoxImageView.hidden = YES;
    }
}

#pragma mark - Table view data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dialogs.count;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    QBChatDialog *chatDialog = self.dialogs[indexPath.row];
    if (self.action != ChatActionDelete || chatDialog.type == QBChatDialogTypePublicGroup) {
        return NO;
    }
    return YES;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self handleSelectRowAtIndexPath:indexPath];
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self handleSelectRowAtIndexPath:indexPath];
}

- (void)handleSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    QBChatDialog *dialog = self.dialogs[indexPath.row];
    if (self.action == ChatActionDelete && dialog.type == QBChatDialogTypePublicGroup) {
        NSString *title = [NSString stringWithFormat:@"You cannot leave %@", dialog.name ];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self showAlertWithTitle:title message:nil fromViewController:self handler:nil];
        });
    } else {
        if ([self.selectedPaths containsObject:indexPath]) {
            [self.selectedPaths removeObject:indexPath];
        } else {
            [self.selectedPaths addObject:indexPath];
        }
        [self checkNavRightBarButtonEnabled];
        [self setupNavigationTitle];
        [self.tableView reloadData];
    }
}

@end

File: sample-chat-swift/sample-chat-swift/Components/DialogsScreen/DialogsSelectionViewController.swift
//
//  DialogsSelectionVC.swift
//  sample-chat-swift
//
//  Created by Injoit on 10/11/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit
import Quickblox

struct DialogsSelectionConstant {
    static let forwarded = "Message forwarded"
}

class DialogsSelectionViewController: DialogListViewController {
    
    //MARK: - Properties
    private var selectedPaths = Set<IndexPath>()
    private var titleView = TitleView()
    var action: ChatAction?
    var message: QBChatMessage?
    internal var senderID: UInt = 0
    
    //MARK: - Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupDialogs()
    }
    
    //MARK - Setup
    override func setupNavigationBar() {
        guard currentUser.isFull == true else {
            return
        }
        senderID = currentUser.ID
        navigationItem.titleView = titleView
        
        let backButtonItem = UIBarButtonItem(image: UIImage(named: "chevron"),
                                             style: .plain,
                                             target: self,
                                             action: #selector(didTapBack(_:)))
        navigationItem.leftBarButtonItem = backButtonItem
        backButtonItem.tintColor = .white
        
        if let action = action {
            if action == .delete {
                let deleteButtonItem = UIBarButtonItem(title: "Delete",
                                                       style: .plain,
                                                       target: self,
                                                       action: #selector(didTapDelete(_:)))
                navigationItem.rightBarButtonItem = deleteButtonItem
                deleteButtonItem.tintColor = .white
            } else if action == .forward {
                let sendButtonItem = UIBarButtonItem(title: "Send",
                                                     style: .plain,
                                                     target: self,
                                                     action: #selector(didTapSend(_:)))
                navigationItem.rightBarButtonItem = sendButtonItem
                sendButtonItem.tintColor = .white
            }
        }
        navigationItem.rightBarButtonItem?.isEnabled = false
    }
    
    override func setupNavigationTitle() {
        let title = action == .delete ? DialogsConstant.deleteChats : DialogsConstant.forward
        let chats = selectedPaths.count == 1 ? "chat" : "chats"
        let numberChats = "\(selectedPaths.count) \(chats) selected"
        titleView.setupTitleView(title: title, subTitle: numberChats)
    }
    
    override func setupDialogs() {
        reloadContent()
    }
    
    //MARK: - Actions
    private func checkNavRightBarButtonEnabled() {
        navigationItem.rightBarButtonItem?.isEnabled = selectedPaths.isEmpty == true ? false : true
    }
    
    @objc func didTapSend(_ sender: UIBarButtonItem) {
        if QBChat.instance.isConnected == false {
            showNoInternetAlert(handler: nil)
            return
        }
        guard let originMessage = message else {
            return
        }
        sender.isEnabled = false
        let sendGroup = DispatchGroup()
        for indexPath in self.selectedPaths {
            let dialog = self.dialogs[indexPath.row]
            guard let dialogID = dialog.id else {
                continue
            }
            sendGroup.enter()
            let forwardedMessage = QBChatMessage.markable()
            forwardedMessage.senderID = senderID
            forwardedMessage.deliveredIDs = [(NSNumber(value: senderID))]
            forwardedMessage.readIDs = [(NSNumber(value: senderID))]
            forwardedMessage.dateSent = Date()
            forwardedMessage.customParameters[Key.saveToHistory] = true
            let originSenderUser = chatManager.storage.user(withID: originMessage.senderID)
            if let fullName = originSenderUser?.fullName {
                forwardedMessage.customParameters[Key.forwardedMessage] = fullName
            } else {
                let currentUser = Profile()
                forwardedMessage.customParameters[Key.forwardedMessage] = currentUser.fullName
            }
            forwardedMessage.dialogID = dialogID
            if let attachment = originMessage.attachments?.first {
                forwardedMessage.text = ChatManagerConstant.attachment
                forwardedMessage.attachments = [attachment]
            } else {
                forwardedMessage.text = originMessage.text
            }
            chatManager.send(forwardedMessage, to: dialog) { [weak self] (error) in
                sendGroup.leave()
                if let error = error {
                    debugPrint("[DialogsSelectionVC] sendMessage error: \(error.localizedDescription)")
                }
                self?.selectedPaths.remove(indexPath)
            }
        }
        sendGroup.notify(queue: DispatchQueue.main) {
            self.navigationController?.popViewController(animated: false)
        }
    }
    
    @objc func didTapDelete(_ sender: UIBarButtonItem) {
        if QBChat.instance.isConnected == false {
            showNoInternetAlert(handler: nil)
            return
        }
        let baseAlertMessage = "Do you really want to leave selected dialog"
        let alertMessage = self.selectedPaths.count == 1 ? baseAlertMessage + "?" : baseAlertMessage + "s?"
        let alertController = UIAlertController(title: "Warning",
                                                message: alertMessage,
                                                preferredStyle: .alert)
        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        let leaveAction = UIAlertAction(title: "Leave", style: .default) { (action:UIAlertAction) in
            sender.isEnabled = false
            let deleteGroup = DispatchGroup()
            for indexPath in self.selectedPaths {
                let dialog = self.dialogs[indexPath.row]
                self.progressView.start()
                guard let dialogID = dialog.id, dialog.type != .publicGroup else {
                    continue
                }
                deleteGroup.enter()
                self.chatManager.leaveDialog(withID: dialogID) { [weak self] error in
                    deleteGroup.leave()
                    if error == nil {
                        self?.selectedPaths.remove(indexPath)
                    }
                }
            }
            deleteGroup.notify(queue: DispatchQueue.main) {
                self.handleLeaveDialog()
            }
        }
        alertController.addAction(cancelAction)
        alertController.addAction(leaveAction)
        present(alertController, animated: true, completion: nil)
    }
    
    @objc func didTapBack(_ sender: UIBarButtonItem) {
        navigationController?.popViewController(animated: false)
    }
    
    // MARK: - UITableViewDataSource
    override func configure(cell: DialogCell, for indexPath: IndexPath) {
        tableView.allowsMultipleSelection = true
        cell.checkBoxView.isHidden = false
        cell.lastMessageDateLabel.isHidden = true
        
        if self.selectedPaths.contains(indexPath) {
            cell.contentView.backgroundColor = UIColor(red:0.85, green:0.89, blue:0.97, alpha:1)
            cell.checkBoxImageView.isHidden = false
            cell.checkBoxView.setRoundBorderEdgeColorView(cornerRadius: 4.0,
                                                          borderWidth: 0.0,
                                                          color: UIColor(red:0.22, green:0.47, blue:0.99, alpha:1),
                                                          borderColor: UIColor(red:0.22, green:0.47, blue:0.99, alpha:1))
        } else {
            cell.contentView.backgroundColor = .clear
            cell.checkBoxView.backgroundColor = .clear
            cell.checkBoxView.setRoundBorderEdgeColorView(cornerRadius: 4.0,
                                                          borderWidth: 1.0,
                                                          borderColor: UIColor(red:0.42, green:0.48, blue:0.57, alpha:1))
            cell.checkBoxImageView.isHidden = true
        }
    }
    
    // MARK: - UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        handleSelectRowAtIndexPath(indexPath)
    }
    
    override func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        handleSelectRowAtIndexPath(indexPath)
    }
    
    private func handleSelectRowAtIndexPath(_ indexPath: IndexPath) {
        let dialog = dialogs[indexPath.row]
        if let action = action, action == .delete,
            dialog.type == .publicGroup {
            DispatchQueue.main.async {
                let alertController = UIAlertController(title: "You cannot leave \(dialog.name ?? "Public chat")",
                                                        message: nil,
                                                        preferredStyle: .alert)
                
                let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
                alertController.addAction(cancelAction)
                self.present(alertController, animated: true, completion: nil)
            }
        } else {
            if selectedPaths.contains(indexPath) {
                selectedPaths.remove(indexPath)
            } else {
                selectedPaths.insert(indexPath)
            }
            checkNavRightBarButtonEnabled()
            setupNavigationTitle()
            tableView.reloadData()
        }
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        let dialog = dialogs[indexPath.row]
        if let action = action, action != .delete || dialog.type == .publicGroup {
            return false
        }
        return true
    }
    
    // MARK: - Helpers
    override func reloadContent() {
        dialogs = chatManager.storage.dialogsSortByUpdatedAt()
        setupNavigationTitle()
        tableView.reloadData()
    }
    
    override func handleLeaveDialog() {
        super.handleLeaveDialog()
        self.navigationController?.popViewController(animated: false)
    }
}
